### PR TITLE
add granular allow list and multiple hooks and fix revert tests

### PR DESCRIPTION
**Pull request by ericneil-sanc**
# Enhanced Permission System with Comprehensive Test Coverage

## Summary

This PR extends the allowlist system with comprehensive test coverage for all permission flag values and combinations, ensuring robust access control for both liquidity and swap operations in the permissioned pools system.

### 1. Extended PermissionFlag Library
- Added bitwise operators (`|`, `&`, `==`) to `PermissionFlag` type for cleaner permission combinations
- Added helper functions (`or`, `and`, `eq`, `hasFlag`) for permission checking
- Fixed operator definitions to use global scope with proper Solidity syntax

### 2. Comprehensive Permission Flag Testing
- Added extensive test coverage for all permission flag values in both `PermissionedPositionManager.t.sol` and `PermissionedV4Router.t.sol`
- Tests cover: `NONE`, `SWAP_ALLOWED`, `LIQUIDITY_ALLOWED`, `ALL_ALLOWED`
- Added combination tests (bitwise OR operations)
- Added edge case tests (zero permissions, max permissions)
- Added dynamic permission change tests
- Added cross-pool type testing

### 3. Test Coverage Includes
- **Position Manager**: Liquidity operations with various permission combinations
- **Router**: Swap operations with different permission levels
- **Edge Cases**: Partial permissions, permission transitions, invalid combinations
- **Multi-hop Swaps**: Permission validation across complex swap paths
- **Exact Input/Output**: Both swap types with permission enforcement

## Technical Details

- **Permission System**: Uses `bytes2` bitwise flags for efficient permission storage and checking
- **Constants**: `SWAP_ALLOWED = 0x0001`, `LIQUIDITY_ALLOWED = 0x0002`, `ALL_ALLOWED = 0xFFFF`
- **Compatibility**: Maintains backward compatibility with existing permission checks
- **Gas Efficiency**: Bitwise operations provide efficient permission validation

## Testing Results

- **18 tests pass** across both test files
- Comprehensive coverage of permission flag combinations and edge cases

## Files Addes

-   `src/hooks/permissionedPools/libraries/PermissionFlags.sol`

## Files Modified

-   `src/hooks/permissionedPools/BaseAllowListChecker.sol`
- `src/hooks/permissionedPools/PermissionedHooks.sol`
-   `src/hooks/permissionedPools/PermissionedPositionManager.sol`
- `src/hooks/permissionedPools/PermissionedV4Router.sol`
-    `src/hooks/permissionedPools/WrappedPermissionedToken.sol`
-   `src/hooks/permissionedPools/interfaces/IAllowlistChecker.sol`
-    `src/hooks/permissionedPools/interfaces/IWrappedPermissionedToken.sol`
-   `test/hooks/permissionedPools/PermissionedPoolsBase.sol`
-   `test/hooks/permissionedPools/PermissionedPositionManager.t.sol`
-   `test/hooks/permissionedPools/PermissionedV4Router.t.sol`
-   `test/hooks/permissionedPools/WrappedPermissionedToken.t.sol`
-   `test/hooks/permissionedPools/WrappedPermissionedTokenFactory.t.sol`
-  `test/hooks/permissionedPools/shared/PermissionedDeployers.sol`
-   `test/hooks/permissionedPools/shared/PermissionedPosmTestSetup.sol`
-  `test/hooks/permissionedPools/shared/PermissionedRoutingTestHelpers.sol`
-   `test/shared/Deploy.sol`

This enhancement provides a flexible permission system with thorough test coverage to ensure proper access control for both liquidity and swap operations in the permissioned pools system.

### CODE REVIEW

File: src/hooks/permissionedPools/PermissionedPositionManager.sol [hunk 1]
```diff
@@ -70,10 +71,33 @@ contract PermissionedPositionManager is PositionManager {
...
     }
 
+    function _checkAllowedHooks(PoolKey calldata poolKey) internal view returns (bool) {
+        if (
+            !_checkAllowedHook(poolKey.currency0, poolKey.hooks) || !_checkAllowedHook(poolKey.currency1, poolKey.hooks)
+        ) {
+            return false;
+        }
+        return true;
+    }
```
Comment (gretzke) [inline src/hooks/permissionedPools/PermissionedPositionManager.sol:86]:
```suggestion
    function _checkAllowedHooks(PoolKey calldata key) internal view returns (bool) {
        return _checkAllowedHook(key.currency0, key.hooks) && _checkAllowedHook(key.currency1, key.hooks);
    }
```
logic can be simplified

File: src/hooks/permissionedPools/PermissionedPositionManager.sol [hunk 2]
```diff
@@ -70,10 +71,33 @@ contract PermissionedPositionManager is PositionManager {
...
+
+    function _checkAllowedHook(Currency currency, IHooks hooks) internal view returns (bool) {
+        address permissionedToken = _verifiedPermissionedTokenOf(currency);
+        if (permissionedToken == address(0)) return true;
+        return isAllowedHooks[currency][hooks];
+    }
+
+    /// @dev Sets which hooks are allowed to be used with a wrapped permissioned token. Only callable by the owner of the wrapped permissioned token
+    function setAllowedHook(Currency currency, IHooks hooks, bool allowed) public {
```
Comment (gretzke) [inline src/hooks/permissionedPools/PermissionedPositionManager.sol:95]: function order should follow the solidity style guide:
- external
- public
- internal
- private

also the function should be external

File: src/hooks/permissionedPools/PermissionedPositionManager.sol [hunk 3]
```diff
@@ -70,10 +71,33 @@ contract PermissionedPositionManager is PositionManager {
...
+        return true;
+    }
+
+    function _checkAllowedHook(Currency currency, IHooks hooks) internal view returns (bool) {
+        address permissionedToken = _verifiedPermissionedTokenOf(currency);
+        if (permissionedToken == address(0)) return true;
+        return isAllowedHooks[currency][hooks];
+    }
+
+    /// @dev Sets which hooks are allowed to be used with a wrapped permissioned token. Only callable by the owner of the wrapped permissioned token
```
Comment (gretzke) [inline src/hooks/permissionedPools/PermissionedPositionManager.sol:94]: external function: add proper natspec with `@notice`, `@param`, etc

**Review by gretzke — CHANGES_REQUESTED** [review]

**Review by gretzke — APPROVED** [review]
